lighthouse-core/report/html/components/gauge.js

```diff
@@ -0,0 +1,398 @@
+  height: var(--plugin-badge-size);
+  background-color: var(--plugin-badge-background-color);
+  background-image: var(--plugin-icon-url);
+  background-repeat: no-repeat;
+  background-size: var(--plugin-icon-size);
+  background-position: 58% 50%;
+  content: "";
+  position: absolute;
+  right: -6px;
+  bottom: 0px;
+  display: block;
+  z-index: 100;
+  box-shadow: 0 0 4px rgba(0,0,0,.2);
+  border-radius: 25%;
+}
+.lh-category .lh-gauge__wrapper--plugin .lh-gauge__svg-wrapper::before {
+  width: var(--plugin-badge-size-big);
+  height: var(--plugin-badge-size-big);
+}
+
+@keyframes load-gauge {
+  from { stroke-dasharray: 0 352; }
+}
+
+.lh-gauge__percentage {
+  width: 100%;
+  height: var(--gauge-circle-size);
+  position: absolute;
+  font-family: var(--report-font-family-monospace);
+  font-size: calc(var(--gauge-circle-size) * 0.34 + 1.3px);
+  line-height: 0;
+  text-align: center;
+  top: calc(var(--score-container-padding) + var(--gauge-circle-size) / 2);
+}
+
+.lh-category .lh-gauge__percentage {
+  --gauge-circle-size: var(--gauge-circle-size-big);
+  --gauge-percentage-font-size: var(--gauge-percentage-font-size-big);
+}
+
+.lh-gauge__wrapper {
+  position: relative;
+  display: flex;
+  align-items: center;
+  flex-direction: column;
+  text-decoration: none;
+  padding: var(--score-container-padding);
+
+  --circle-border-width: 8;
+  --transition-length: 1s;
+
+  /* Contain the layout style paint & layers during animation*/
+  contain: content;
+  will-change: opacity; /* Only using for layer promotion */
+}
+
+.lh-gauge__label {
+  font-size: var(--gauge-label-font-size);
+  line-height: var(--gauge-label-line-height);
+  margin-top: 10px;
+  text-align: center;
+  color: var(--report-text-color);
+}
+
+/* TODO(#8185) use more BEM (.lh-gauge__label--big) instead of relying on descendant selector */
+.lh-category .lh-gauge__label {
+  --gauge-label-font-size: var(--gauge-label-font-size-big);
+  --gauge-label-line-height: var(--gauge-label-line-height-big);
+  margin-top: 14px;
+}
+  </style>
+
+  <a href="#" class="lh-gauge__wrapper">
+    <!-- Wrapper exists for the ::before plugin icon. Cannot create pseudo-elements on svgs. -->
+    <div class="lh-gauge__svg-wrapper">
+      <svg viewBox="0 0 120 120" class="lh-gauge">
+        <circle class="lh-gauge-base" r="56" cx="60" cy="60"></circle>
+        <circle class="lh-gauge-arc" transform="rotate(-90 60 60)" r="56" cx="60" cy="60"></circle>
+      </svg>
+    </div>
+    <div class="lh-gauge__percentage"></div>
+    <div class="lh-gauge__label"></div>
+  </a>
+`;
+
+const attributes = {
+  score: '0.5',
+  variant: 'pass',
+  label: 'Lighthouse',
+};
+
+class LighthouseGaugeElement extends HTMLElement {
+  constructor() {
+    super();
+    this._shadowRoot = this.attachShadow({mode: 'open'});
+    this._shadowRoot.appendChild(template.content.cloneNode(true));
+    this._attributes = {...attributes};
+
+    for (const key in Object.keys(attributes)) {
```

Comment:
This chunk from 319-354 feels like it will be copy pasted in every web component we make

Reply:
deferring a base class until a second component appears?

Reply:
Oh forgot to add this to the PR hang on :) 

basically it's unclear how we could continue with our existing strategy for the report without bundling and possibly share code between components.

Reply:
> without bundling

well, you know my feeling on the matter (#8632). I am all for more build-y things in the report.

Could we have a "variables" component (which you must include in any page that uses LH components), and remove all (shared) variables from each individual component?

Reply:
the challenge here is that the HTML report needs to be a standalone file, we "bundle" in a sense for the report today by concating all the files into a single script tag that then use global references to their dependencies.

this obviously poses problems for anyone who wants to reuse a class then in another project that is using an _actual_ bundler

Reply:
there's not a way to create a script tag and fake the source URL to the browser `import('./my-component')` would work now is there?

@connorjclark does the magic of `//# sourceURL` extend that far? 😆

Reply:
Would you be able to use Rollup to produce that bundle? This would let you use modules.

`data:` URLs might work, but that would be highly unusual and hard for regular JS devs to understand, IMO.

Reply:
> there's not a way to create an inline script and fake the source URL to the browser so `import('./my-component')` would work now is there?
> 
> @connorjclark does the magic of `//# sourceURL` extend that far?

no... sourceURL is not an ES feature :) only browser developer tooling uses it.